src/java_bytecode/java_types.h

```diff
@@ -518,6 +518,82 @@ to_java_specialized_generic_class_type(typet &type)
   return static_cast<const java_specialized_generic_class_typet &>(type);
 }
 
+class java_symbol_typet : public symbol_typet
+{
+public:
+  java_symbol_typet() : symbol_typet()
+  {
+  }
+
+  explicit java_symbol_typet(const irep_idt &identifier)
+    : symbol_typet(identifier)
+  {
+  }
+};
+
+inline const java_symbol_typet &to_java_symbol_type(const typet &type)
+{
+  PRECONDITION(type.id() == ID_symbol);
+  return static_cast<const java_symbol_typet &>(type);
+}
+
+inline java_symbol_typet &to_java_symbol_type(typet &type)
+{
+  PRECONDITION(type.id() == ID_symbol);
+  return static_cast<java_symbol_typet &>(type);
+}
+
+/// Symbol for generics, extends the java symbol type with a
+/// vector of java generic type parameters (also called type variables). This
+/// can be used for example to store generic superclass symbol.
+class java_generic_symbol_typet : public java_symbol_typet
+{
+public:
+  typedef std::vector<reference_typet> generic_typest;
+
+  java_generic_symbol_typet(
+    const java_symbol_typet &symbol_type,
+    const generic_typest &gen_types)
+    : java_symbol_typet(symbol_type)
+  {
+    set(ID_C_java_generic_symbol_type, true);
+    for(const auto &type : gen_types)
+    {
+      generic_types().push_back(type);
+    }
+  }
+
+  const generic_typest &generic_types() const
+  {
+    return (const generic_typest &)(find(ID_generic_types).get_sub());
+  }
+
+  generic_typest &generic_types()
+  {
+    return (generic_typest &)(add(ID_generic_types).get_sub());
```

Comment:
I was confused by the name of the function `add`, but yes that this is correct...